### PR TITLE
Create Automation for API Gateway

### LINKED ISSUE #56: Create automation for API gateway
Use Case
As a developer, I want to write a script that creates an API gateway so I don't have to manually do it myself.

Success Criteria 
- [ ] Create script that will auto generate an API gateway
- [ ] the API gateway paths are properly configured as described in the Swagger document located in: 

/development/backend/apis/N5251-AIFeedbackCoach-1.0-resolved.yaml

Time to complete
Estimated: 6 hours
Actual:

### PR DESCRIPTION
**Pull request by Jflands23**
## Description
As a developer, I want to write a script that creates an API gateway, so I don't have to manually do it myself.
Closes #56 
## Testing Steps
**Prerequisite**
•	  Must have [AWS CLI](https://aws.amazon.com/cli/) set up with your credentials

**Creating the API Gateway that contains a Mock API**
- [ ]	  Open a terminal window running bash
- [ ]	  Navigate to the makeAPIGateway folder
- [ ]	  Run ./makeAPIGateway.sh
- [ ]	  Follow the prompts to enter your first and last initial (image 1)
- [ ]	  View the automatically generated names for the API Gateway and Mock API
- [ ]	  Enter y to confirm creation of REST API and API Gateway, or enter n to exit the program
- [ ]	  If y is entered, an API Gateway containing Mock REST API will be deployed
- [ ]	  After stack has been created, Navigate to the AWS console 
- [ ]	  Search API Gateway to view created APIs, and confirm newly created Mock REST API is present (image 2)

## Images
1.	The script in the terminal
PRP1

2.	The generated mock api and api gateway on AWS console
PRP2

## Known Issues
This script is a shell script, and will not work if you are not using bash in your terminal. If you are using powershell, you can configure your computer to run bash on Windows.

### CODE REVIEW

File: devops/makeAPIGateway/makeAPIGateway.sh [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      done
+
+    if [ "$initials" == "q" ]; then
+      printf '%s\n' "Done"
+      break
+    fi
+
+
+    # converts initials to lowercase and remove spaces
+    formattedInitials="$(echo "${initials}" | tr -d '[:space:]' | tr '[:upper:]' '[:lower:]')"
```
Comment (dansjack) [inline devops/makeAPIGateway/makeAPIGateway.sh:32]: Heads up: if this script ends up being ran through a github action, this way of making a random string wont work. You can instead use `openssl`:

```bash
OPENSSL_STRING=$(openssl rand -hex 3)
RANDOM_STRING=${OPENSSL_STRING:0:5}
```

See the file in .github/scripts in #69 for an example

File: devops/makeAPIGateway/makeAPIGateway.yml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+  ApiGatewayRestApi:
+    Type: AWS::ApiGateway::RestApi
+    Properties:
+      ApiKeySourceType: HEADER
+      Description: An API Gateway with a Mock Integration
+      EndpointConfiguration:
+        Types:
+          - REGIONAL
+      Name: mock-api
```
Comment (dansjack) [inline devops/makeAPIGateway/makeAPIGateway.yml:15]: Just checking, is it intentional that the API name is always the same? 

The script shows you what the API name would be using our naming convention, but no name ref is passed to this template in order to use it

image

**Review by dansjack — COMMENTED** [review]
I left a comment and had a question, but the script is working!

**Review by KristinJJ — CHANGES_REQUESTED** [review]
Hi @Jflands23, the automation works!

However I would like to request that the automation creates a unique name for each API Gateway upon creation (similar to how the CloudFormation that is created has a unique name), otherwise all the API Gateway created with this script are named `mock-api`, which makes it difficult to see which API Gateway is created when this script is run. 

Screen Shot 2022-02-15 at 6 56 55 PM

File: devops/makeAPIGateway/makeAPIGateway.sh (hunk 1 above)
Comment (Jflands23) [reply devops/makeAPIGateway/makeAPIGateway.sh:32]: Thanks for the feedback! made these changes so there shouldn't be problems in the future

File: devops/makeAPIGateway/makeAPIGateway.yml (hunk 2 above)
Comment (Jflands23) [reply devops/makeAPIGateway/makeAPIGateway.yml]: I made the names unique now to cause less confusion

**Comment by Jflands23** [comment]
> Hi @Jflands23, the automation works!
> 
> However I would like to request that the automation creates a unique name for each API Gateway upon creation (similar to how the CloudFormation that is created has a unique name), otherwise all the API Gateway created with this script are named `mock-api`, which makes it difficult to see which API Gateway is created when this script is run.
> 
> Screen Shot 2022-02-15 at 6 56 55 PM

Made the changes so now every API has a unique name. Let me know if there are any issues

**Review by KristinJJ — APPROVED** [review]

**Comment by KristinJJ** [comment]
> 
Nice! I can see the API Gateway with the unique name, changes approved. 
Screen Shot 2022-02-20 at 7 08 22 PM

One very minor nit-picky detail I noticed is that the name of the API isn't shown in the terminal (unlike the stack name)
Screen Shot 2022-02-20 at 7 08 04 PM

**Review by dansjack — APPROVED** [review]
Approving with the same note as @KristinJJ, should be an easy fix!

**Pull request merged**